# Data Cleaning

### Dataset: Telco_customer_churn.csv

In [13]:
import numpy as np
import pandas as pd

In [14]:
df = pd.read_csv("Telco_customer_churn.csv")

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   object 
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   object 
 3   State              7043 non-null   object 
 4   City               7043 non-null   object 
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   object 
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   object 
 10  Senior Citizen     7043 non-null   object 
 11  Partner            7043 non-null   object 
 12  Dependents         7043 non-null   object 
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   object 
 15  Multiple Lines     7043 non-null   object 
 16  Internet Service   7043 

In [16]:
df.dtypes

CustomerID            object
Count                  int64
Country               object
State                 object
City                  object
Zip Code               int64
Lat Long              object
Latitude             float64
Longitude            float64
Gender                object
Senior Citizen        object
Partner               object
Dependents            object
Tenure Months          int64
Phone Service         object
Multiple Lines        object
Internet Service      object
Online Security       object
Online Backup         object
Device Protection     object
Tech Support          object
Streaming TV          object
Streaming Movies      object
Contract              object
Paperless Billing     object
Payment Method        object
Monthly Charges      float64
Total Charges         object
Churn Label           object
Churn Value            int64
Churn Score            int64
CLTV                   int64
Churn Reason          object
dtype: object

In [17]:
duplicate_rows = df.duplicated().sum()
print("\nNumber of Duplicate Rows:", duplicate_rows)

duplicates = df[df.duplicated()]
print("\nDuplicate Rows Preview:")
print(duplicates.head())


Number of Duplicate Rows: 0

Duplicate Rows Preview:
Empty DataFrame
Columns: [CustomerID, Count, Country, State, City, Zip Code, Lat Long, Latitude, Longitude, Gender, Senior Citizen, Partner, Dependents, Tenure Months, Phone Service, Multiple Lines, Internet Service, Online Security, Online Backup, Device Protection, Tech Support, Streaming TV, Streaming Movies, Contract, Paperless Billing, Payment Method, Monthly Charges, Total Charges, Churn Label, Churn Value, Churn Score, CLTV, Churn Reason]
Index: []

[0 rows x 33 columns]


In [18]:
missing_values = df.isnull().sum()
print("Missing Values in Each Column:")
print(missing_values)

print("\nTotal Missing Values:", df.isnull().sum().sum())

Missing Values in Each Column:
CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

Total Missing Values: 5174


Churn Reason is only filled for customers who actually churned.
Customers who did not churn naturally have no reason, so the value is empty.

From the dataset:

Total customers = 7043

Missing churn reason = 5174

Customers who churned = 7043 − 5174 = 1869
Meaning:

1869 customers churned

5174 customers stayed

This matches the typical churn distribution of the dataset.

For analysis, Replacing missing values with "No Churn":

In [19]:
df['Churn Reason'] = df['Churn Reason'].fillna('No Churn')

Datatype Correction

In [20]:
df['Total Charges'] = pd.to_numeric(df['Total Charges'], errors='coerce')

In [21]:
df['Total Charges'].dtype

dtype('float64')

Consistency Detection

In [22]:
for col in df.columns:
    print(f"\nColumn: {col}")
    print(df[col].unique())


Column: CustomerID
['3668-QPYBK' '9237-HQITU' '9305-CDSKC' ... '2234-XADUH' '4801-JZAZL'
 '3186-AJIEK']

Column: Count
[1]

Column: Country
['United States']

Column: State
['California']

Column: City
['Los Angeles' 'Beverly Hills' 'Huntington Park' ... 'Standish' 'Tulelake'
 'Olympic Valley']

Column: Zip Code
[90003 90005 90006 ... 96128 96134 96146]

Column: Lat Long
['33.964131, -118.272783' '34.059281, -118.30742' '34.048013, -118.293953'
 ... '40.346634, -120.386422' '41.813521, -121.492666'
 '39.191797, -120.212401']

Column: Latitude
[33.964131 34.059281 34.048013 ... 40.346634 41.813521 39.191797]

Column: Longitude
[-118.272783 -118.30742  -118.293953 ... -120.386422 -121.492666
 -120.212401]

Column: Gender
['Male' 'Female']

Column: Senior Citizen
['No' 'Yes']

Column: Partner
['No' 'Yes']

Column: Dependents
['No' 'Yes']

Column: Tenure Months
[ 2  8 28 49 10  1 47 17  5 34 11 15 18  9  7 12 25 68 55 37  3 27 20  4
 58 53 13  6 19 59 16 52 24 32 38 54 43 63 21 69 22 61 6

Multiple Lines Values:
Yes
No
No phone service

Meaning:Customers without phone service cannot have multiple lines.

Correct data logic.

Several Internet-related columns

Columns: Online Security, Online Backup, Device Protection, Tech Support,Streaming TV, Streaming Movies

Values:
Yes
No
No internet service

Meaning:
If Internet Service = No, these services cannot exist.

So "No internet service" is valid and expected.

No cleaning needed.

In [23]:
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1,89,5340,Competitor had better devices


Exporting Cleaned Dataset

In [24]:
df.to_csv("Telco_customer_churn_data.csv", index=False)

##### The Dataset is ready for Dashboard Analysis!